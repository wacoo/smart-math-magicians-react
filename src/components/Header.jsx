import { NavLink } from 'react-router-dom';

const Header = () => (
  <div className="header">
    <h1>Math Magicians</h1>
    <ul>
      <li>
        <NavLink to="/" activeClassName="active">Home</NavLink>
      </li>
      <li>
        <NavLink to="/calculator" activeClassName="active">Calculator</NavLink>
      </li>
      <li>
        <NavLink to="/quotes" activeClassName="active">Quotes</NavLink>
      </li>
    </ul>
  </div>
);

export default Header;
