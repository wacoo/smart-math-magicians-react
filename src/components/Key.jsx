import React from 'react';
import PropTypes from 'prop-types';

const Key = (props) => {
  const { lbl, cls, onSmash } = props;
  return (
    <button type="button" className={cls} onClick={onSmash}>
      {lbl}
    </button>
  );
};

Key.propTypes = { lbl: PropTypes.string.isRequired };
Key.propTypes = { cls: PropTypes.string.isRequired };
Key.propTypes = { onSmash: PropTypes.func.isRequired };
export default Key;
