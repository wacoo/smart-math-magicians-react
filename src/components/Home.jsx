import Header from './Header';

const Home = () => (
  <>
    <Header />
    <div className="container">
      <h2 className="h2">Welcome to our page!</h2>
      <p>
        Smart Math Magicians brings a touch of enchantment to your mathematical
        journey. Equipped with the power of React, this single-page application
        seamlessly melds the art of computation with the wonder of technology.
        With Smart Math Magicians, complex calculations become a breeze.
      </p>
    </div>
  </>
);

export default Home;
