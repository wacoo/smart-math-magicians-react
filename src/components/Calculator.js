import React from 'react';
import Key from './Key';
import Display from './Display';
import calculate from '../logic/calculate';

const selectClass = (idx, divIdx) => {
  if (divIdx < 4 && idx !== 3) {
    return 'normal';
  }
  if (divIdx === 4 && idx === 1) {
    return 'normal';
  }
  if (divIdx === 4 && idx === 0) {
    return 'zero';
  }
  return 'operator';
};

class Calculator extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      obj: {
        total: 0,
        next: 0,
        operation: '',
      },
    };

    this.operation = '';
    this.oprString = '';
  }

  isNumeric = (str) => {
    if (typeof str !== 'string') {
      return false;
    }
    return !Number.isNaN(str) && !Number.isNaN(parseFloat(str));
  }

  handleClick = (btnName) => {
    const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
    const { obj } = this.state;
    this.setState({ obj: calculate(obj, btnName) });
    if (btnName === 'AC') {
      this.setState({ obj: {} });
      this.oprString = '';
    }

    if (btnName === '+/-') {
      const neg = `-${obj.next}`;
      const last = this.oprString.substring(this.oprString.length - 1);
      if (numbers.includes(last) && last !== '0') {
        this.oprString = this.oprString.substring(0, this.oprString.length - 1);
        this.oprString += neg;
      } else {
        this.oprString += '-';
      }

      return;
    }

    if (btnName !== 'AC') {
      if (this.oprString === '=') {
        this.oprString = '=';
      } else {
        this.oprString += btnName;
      }
    }

    this.operation = btnName;
  };

  getDisplayValue = (obj) => {
    const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
    if (this.operation === '=') {
      if (this.oprString === '=') {
        this.oprString = '';
        return 0;
      }

      if (this.isNumeric(this.oprString)) {
        if (obj.total) {
          this.oprString += obj.total;
          return this.oprString;
        }
        return obj.next;
      }

      if (numbers.includes(this.oprString)) {
        return this.oprString;
      }

      this.oprString += obj.total;
      return this.oprString;
    }

    if (this.operation === 'AC') {
      return 0;
    }
    return this.oprString;
  }

  render() {
    const lblArray = [
      ['AC', '+/-', '%', 'รท'],
      ['7', '8', '9', 'x'],
      ['4', '5', '6', '-'],
      ['1', '2', '3', '+'],
      ['0', '.', '='],
    ];
    const { obj } = this.state;
    return (
      <div className="calculator">
        <Display className="input" value={this.getDisplayValue(obj)} />
        {lblArray.map((lbl, divIdx) => (
          <div className="row" key={lbl[0]}>
            {lbl.map((element, idx) => (
              <Key
                lbl={element}
                cls={selectClass(idx, divIdx)}
                key={element}
                on
                onSmash={() => this.handleClick(element)}
              />
            ))}
          </div>
        ))}
      </div>
    );
  }
}

export default Calculator;
