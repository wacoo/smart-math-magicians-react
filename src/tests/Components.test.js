import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import Header from '../components/Header';
import Home from '../components/Home';
import Calculator from '../components/Calculator';
import LoadQuotes from '../components/Quotes';
import Key from '../components/Key';

let init = 1;
function handleClick(val) {
  init = val + 1;
  return val + 1;
}

describe('Test components', () => {
  it('Test Header component', () => {
    const homeComponent = renderer.create(<BrowserRouter><Header /></BrowserRouter>);
    expect(homeComponent).toMatchSnapshot();
  });

  it('Test Home component', () => {
    const headerComponent = renderer.create(<BrowserRouter><Home /></BrowserRouter>);
    expect(headerComponent).toMatchSnapshot();
  });
  it('Test Calculator component', () => {
    const calcComponent = renderer.create(<BrowserRouter><Calculator /></BrowserRouter>);
    expect(calcComponent).toMatchSnapshot();
  });
  it('Test Quote component', () => {
    const quoteComponent = renderer.create(<BrowserRouter><LoadQuotes /></BrowserRouter>);
    expect(quoteComponent).toMatchSnapshot();
  });
});

describe('Test interaction', () => {
  it('Tests Calculator keys button to exist', () => {
    const { getByRole } = render(<Key lbl="Hello" onSmash={() => handleClick(11)} />);
    const key = getByRole('button');
    expect(key).toHaveTextContent('Hello');
  });

  it('Tests Key button click', () => {
    const { getByText } = render(<Key lbl="Hello" onSmash={() => handleClick(12)} />);
    const key = getByText('Hello');
    fireEvent.click(key);

    screen.findByRole('button');
    expect(init).toBe(13);
  });
});
