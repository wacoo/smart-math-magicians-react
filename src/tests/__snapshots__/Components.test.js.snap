// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test components Test Calculator component 1`] = `
Array [
  <div
    className="header"
  >
    <h1>
      Math Magicians
    </h1>
    <ul>
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/calculator"
          onClick={[Function]}
        >
          Calculator
        </a>
      </li>
      <li>
        <a
          href="/quotes"
          onClick={[Function]}
        >
          Quotes
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="container"
  >
    <div
      className="calculator"
    >
      <input
        className="display"
        type="text"
        value=""
      />
      <div
        className="row"
      >
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          %
        </button>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          ÷
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          7
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          8
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          9
        </button>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          x
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          6
        </button>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="zero"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
        <button
          className="normal"
          onClick={[Function]}
          type="button"
        >
          .
        </button>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`Test components Test Header component 1`] = `
Array [
  <div
    className="header"
  >
    <h1>
      Math Magicians
    </h1>
    <ul>
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/calculator"
          onClick={[Function]}
        >
          Calculator
        </a>
      </li>
      <li>
        <a
          href="/quotes"
          onClick={[Function]}
        >
          Quotes
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="container"
  >
    <h2
      className="h2"
    >
      Welcome to our page!
    </h2>
    <p>
      Smart Math Magicians brings a touch of enchantment to your mathematical journey. Equipped with the power of React, this single-page application seamlessly melds the art of computation with the wonder of technology. With Smart Math Magicians, complex calculations become a breeze.
    </p>
  </div>,
]
`;

exports[`Test components Test Home component 1`] = `
<div
  className="header"
>
  <h1>
    Math Magicians
  </h1>
  <ul>
    <li>
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </li>
    <li>
      <a
        href="/calculator"
        onClick={[Function]}
      >
        Calculator
      </a>
    </li>
    <li>
      <a
        href="/quotes"
        onClick={[Function]}
      >
        Quotes
      </a>
    </li>
  </ul>
</div>
`;

exports[`Test components Test Quote component 1`] = `
Array [
  <div
    className="header"
  >
    <h1>
      Math Magicians
    </h1>
    <ul>
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/calculator"
          onClick={[Function]}
        >
          Calculator
        </a>
      </li>
      <li>
        <a
          href="/quotes"
          onClick={[Function]}
        >
          Quotes
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="container"
  >
    <div
      className="quote"
    >
      <blockquote>
        <p>
           undefined 
        </p>
        <p>
          <strong>
            —
             undefined
          </strong>
        </p>
      </blockquote>
    </div>
  </div>,
]
`;
